R/pkg-dplyr.R

```diff
@@ -298,6 +304,108 @@ relocate.SubstraitCompiler <- function(.data, ..., .before = NULL, .after = NULL
   substrait_select(.data, !!!new_mask)
 }
 
+#' @rdname select.SubstraitCompiler
+#' @importFrom dplyr semi_join
+#' @export
+semi_join.SubstraitCompiler <- function(x, y, by = NULL, ...) {
+  rlang::check_dots_empty()
+
+  # Semi join support is limited so just do an inner join emitting only
+  # columns from the left.
+  substrait_join(
+    x, y,
+    by = by,
+    type = "JOIN_TYPE_INNER",
+    output_mapping = function(by, names_left, names_right) seq_along(names_left) - 1L,
+    name_repair = join_name_repair_none
+  )
+}
+
+#' @rdname select.SubstraitCompiler
+#' @importFrom dplyr anti_join
+#' @export
+anti_join.SubstraitCompiler <- function(x, y, by = NULL, ...) {
+  rlang::check_dots_empty()
+
+  joined <- substrait_join(
+    x, y,
+    by = by,
+    type = "JOIN_TYPE_ANTI",
+    output_mapping = join_emit_all,
+    name_repair = join_name_repair_none
+  )
+
+  # A join emit doesn't seem to work with DuckDB yet, so to make this work
+  # we add a project instead of applying the output mapping at the join step
+  select_args <- rlang::set_names(seq_along(x$schema$names), x$schema$names)
+  dplyr::select(joined, !!!select_args)
+}
+
+#' @rdname select.SubstraitCompiler
+#' @importFrom dplyr inner_join
+#' @export
+inner_join.SubstraitCompiler <- function(x, y, by = NULL, suffix = c(".x", ".y"),
+                                         ..., keep = NULL) {
+  dplyr_mutating_join("JOIN_TYPE_INNER", x, y, by = by, suffix = suffix, ..., keep = keep)
+}
+
+#' @rdname select.SubstraitCompiler
+#' @importFrom dplyr left_join
+#' @export
+left_join.SubstraitCompiler <- function(x, y, by = NULL, suffix = c(".x", ".y"),
```

Comment:
This function signature doesn't match dplyr's, which is fine for now as we're not going for a feature complete implementation, but how about we start either including the unimplemented arguments with an error message saying we don't support them yet, or at least open an issue so we can keep track of the work for later?

Reply:
I don't love the error message as it's not indicative of the actual issue, though happy to leave it for now.